### ISSUE TITLE
Detection not working

### ISSUE THREAD

**Issue by wudashuo** [ultralytics/yolov5#1625]
## 🐛 Bug
detect.py not working normally on images and videos, but works well on streaming media(rtsp)

## To Reproduce (REQUIRED)

Input:
```
python detect.py
```

Output:
```
Namespace(agnostic_nms=False, augment=False, classes=None, conf_thres=0.25, device='', exist_ok=False, img_size=640, iou_thres=0.45, name='exp', project='runs/detect', save_conf=False, save_txt=False, source='data/images', update=False, view_img=False, weights='yolov5s.pt')
Using torch 1.7.0+cu110 CUDA:0 (GeForce GTX 1660 Ti, 6144MB)

Fusing layers...
Model Summary: 232 layers, 7459581 parameters, 0 gradients
image 1/2 D:\Documents\Code\yolov5\data\images\bus.jpg: 640x480 Done. (0.031s)
image 2/2 D:\Documents\Code\yolov5\data\images\zidane.jpg: 384x640 Done. (0.016s)
Results saved to runs\detect\exp
```

Output nothing, and the output images are the same as the original images.
And I tried detecting videos, the same problem.

After a while I found it worked fine with streaming video:

Input:
```
python .\detect.py --source rtsp://192.168.6.42:554
```

Output:
```
Namespace(agnostic_nms=False, augment=False, classes=None, conf_thres=0.25, device='', exist_ok=False, img_size=640, iou_thres=0.45, name='exp', project='runs/detect', save_conf=False, save_txt=False, source='rtsp://admin:iaac@1234!@192.168.6.42:554', update=False, view_img=False, weights='yolov5s.pt')
Using torch 1.7.0+cu110 CUDA:0 (GeForce GTX 1660 Ti, 6144MB)

Fusing layers...
Model Summary: 232 layers, 7459581 parameters, 0 gradients
1/1: rtsp://192.168.6.42:554...  success (1920x1080 at 25.00 FPS).

0: 384x640 3 chairs, 5 potted plants, Done. (0.519s)
0: 384x640 2 chairs, 5 potted plants, Done. (0.012s)
0: 384x640 2 chairs, 5 potted plants, Done. (0.013s)
0: 384x640 2 chairs, 5 potted plants, Done. (0.011s)
0: 384x640 2 chairs, 5 potted plants, Done. (0.011s)
0: 384x640 2 chairs, 5 potted plants, Done. (0.011s)
```

I found the following code may cause the problem:
```
if webcam:
        view_img = True
        cudnn.benchmark = True  # set True to speed up constant image size inference
        dataset = LoadStreams(source, img_size=imgsz)
    else:
        save_img = True
        dataset = LoadImages(source, img_size=imgsz)
```

so I tried to add `cudnn.deterministic = True` or `cudnn.benchmark = True`  in the 'else:', and `cudnn.benchmark = True` worked well:
```
if webcam:
        view_img = True
        cudnn.benchmark = True  # set True to speed up constant image size inference
        dataset = LoadStreams(source, img_size=imgsz)
    else:
        save_img = True
        cudnn.benchmark = True # add
        dataset = LoadImages(source, img_size=imgsz)
```
```
PS D:\Documents\Code\yolov5> python .\detect.py
Namespace(agnostic_nms=False, augment=False, classes=None, conf_thres=0.25, device='', exist_ok=False, img_size=640, iou_thres=0.45, name='exp', project='runs/detect', save_conf=False, save_txt=False, source='data/images', update=False, view_img=False, weights='yolov5s.pt')
Using torch 1.7.0+cu110 CUDA:0 (GeForce GTX 1660 Ti, 6144MB)

Fusing layers...
Model Summary: 232 layers, 7459581 parameters, 0 gradients
image 1/2 D:\Documents\Code\yolov5\data\images\bus.jpg: 640x480 4 persons, 1 buss, 1 skateboards, Done. (0.546s)
image 2/2 D:\Documents\Code\yolov5\data\images\zidane.jpg: 384x640 2 persons, 2 ties, Done. (0.532s)
Results saved to runs\detect\exp3
```

## Environment

 - OS: [Windows 10 20H2]
 - GPU: [1660 Ti]
 - Python: [3.8.6]
 - CUDA: [11.0]
 - PyTorch: [1.7.0+cu110]

## Additional context
1.   At first I install pytorch via `pip install torch===1.7.0+cu110 torchvision===0.8.1+cu110 torchaudio===0.7.0 -f https://download.pytorch.org/whl/torch_stable.html` as the pytorch.org instruction. But an error of importing opencv occurred: 
```
  ** On entry to DGEBAL parameter number  3 had an illegal value
 ** On entry to DGEHRD  parameter number  2 had an illegal value
 ** On entry to DORGHR DORGQR parameter number  2 had an illegal value
 ** On entry to DHSEQR parameter number  4 had an illegal value
Traceback (most recent call last):
  File "D:\Softwares\Python\lib\site-packages\numpy\__init__.py", line 305, in <module>
    _win_os_check()
  File "D:\Softwares\Python\lib\site-packages\numpy\__init__.py", line 302, in _win_os_check
    raise RuntimeError(msg.format(__file__)) from None
RuntimeError: The current Numpy installation ('D:\\Softwares\\Python\\lib\\site-packages\\numpy\\__init__.py') fails to pass a sanity check due to a bug in the windows runtime. See this issue for more information: https://tinyurl.com/y3dm3h86
Traceback (most recent call last):
  File ".\detect.py", line 5, in <module>
    import cv2
  File "D:\Softwares\Python\lib\site-packages\cv2\__init__.py", line 5, in <module>
    from .cv2 import *
ImportError: numpy.core.multiarray failed to import
 ```
I found the solution: downgrade numpy to 1.19.3, that may help anyone who encounter this problem.

2. I also found two small bugs, 
(a) if detecting streaming media, `python .\detect.py --source rtsp://192.168.6.42:554`, there will be an empty folder created. 
(b) if `python .\detect.py --source rtsp://192.168.6.42:554 --save-txt`, there is only a 'admin' file in exp/labels folder.

**Comment by wudashuo** [comment]
Forget to say: Thx for your amazing work and constantly updating, glenn！

Additional: This is a new PC that I just bought. At first I install Python3.9.0 + CUDA11.1, the problem occurred, so I downgrade to Python3.8.6 + CUDA11.0, and still the same.
Interestingly, my another PC was Python3.7.8(you said 3.8 or later) + CUDA11.0 + Pytorch1.6.0+cu101(not exactly compatible with cuda version) , it works fine and everything is normal!
Now that CUDA11.0 on my another PC works well, maybe it's the Pytorch version(1.7+cu110) that caused the problem?

**Comment by glenn-jocher** [comment]
@wudashuo thanks for the info. If you had to summarize, it appears that the main actionable insight from your experiments is that cudnn.benchmark appears to have an effect on inference results?

Your other two observations, yes an empty directory is created on every inference session, to be populated by outputs depending on the arguments supplied, such as --save-txt. 

You also say --save-txt is not functioning with streaming sources?

**Comment by wudashuo** [comment]
> @wudashuo thanks for the info. If you had to summarize, it appears that the main actionable insight from your experiments is that cudnn.benchmark appears to have an effect on inference results?
> 
> Your other two observations, yes an empty directory is created on every inference session, to be populated by outputs depending on the arguments supplied, such as --save-txt.
> 
> You also say --save-txt is not functioning with streaming sources?

**1**. I am pretty sure that pytorch version that cause the problem. I don't know it's everyone or just me, pytorch 1.7.0 and 1.7.1 won't work on Windows, detecting nothing at all, unless set `cudnn.benchmark = True`, then the detection is normal.  
![image](https://user-images.githubusercontent.com/5948604/101873291-a6513980-3bc1-11eb-9ba8-2808d5ffdbbd.png)
![image](https://user-images.githubusercontent.com/5948604/101873421-def11300-3bc1-11eb-9035-330e7ba4ebbc.png)

But Pytorch 1.7 on Linux works well. 
![image](https://user-images.githubusercontent.com/5948604/101873621-3c855f80-3bc2-11eb-9aef-1e4df5e9a5bf.png)

**To summarize:**
(1) Windows + pytorch 1.7.0 or 1.7.1: NOT working.
(2) Windows + pytorch 1.7.0 or 1.7.1, set `cudnn.benchmark = True` : working.
(3) Linux + pytorch 1.7.0: working.
(4) Windows + pytorch 1.6.0: working.




**2**. Yes, `--save-txt` is not functioning with streaming sources,  it will generate a empty file in `exp/labels` folder instead of labels. Because `p` or `p.stem` or `p.name` in this situation always be the rtsp link, not path, so when opening txt_path to write results, there will be an error.
https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/detect.py#L88-L89 https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/detect.py#L106

BTW: 
1. On Windows, there is no suitable torch or torchvision version to be installed through pypi,  it only can be installed from pytorch.org, if using `pip install -r requirements.txt`, there will be an error.
2.  I found in some cases there will be no problem even if your torch version is incompatible with your CUDA version, like CUDA 11.0 or 11.1 is compatible with torch 1.6.0+CU101.
3. When trying to solve the empty label problem,  I wonder it's hard to name the label txt files with streaming sources detection, comparing to named by image names or video frames. Perhaps using current time(yyyy-MM-dd-HH:mm:ss.SSS) is a solution?

**Comment by glenn-jocher** [comment]
Ok got it, thanks for the details. We aim to support all 3 major operating systems, but Windows has always had a complicated relationship with ML and PyTorch especially. I assume these kinks should work themselves out at somepoint going forward, but in general I would recommend Linux for all ML tasks. It works better, and its free.

We have CI tests that run on windows, the last scheduled test was about 14 hours ago and is running on a 1.7.1 environment on CPU, and detections operate correctly:

https://github.com/ultralytics/yolov5/runs/1534297671?check_suite_focus=true
<img width="1560" alt="Screen Shot 2020-12-11 at 6 42 54 AM" src="https://user-images.githubusercontent.com/26833433/101916922-72075a00-3b7c-11eb-8d00-b2df44617267.png">

1. I suppose a simple workaround would be to set `cudnn.benchmark = True` for all cases. I will run some tests to see if there are any adverse effects from this.
2. I will create a TODO for the streaming --save-txt bug. It may just need some special handling for the name as you suggest. We need a naming convention that uniquely identifies a source and a frame index, as we already do for videos, so a timestamp wont work, but im sure theres something we can figure out.

EDIT: I forgot to mention in 2, we need this for the multi-stream detection case, where detect.py runs multiple streams with batched inference.

**Comment by glenn-jocher** [comment]
TODO: fix --save-txt streaming source bug above in detect.py, investigate cudnn.benchmark = True for all detect.py.

**Comment by glenn-jocher** [comment]
I tested `cudnn.benchmark = True` for detect.py, and the results are pretty disastrous. This setting appears to optimize for a specific inference size, with some initial tests before starting, like a JIT compiler would do. The result is that each new shape takes a long time to optimize, so rapidly changing input shapes, like detect.py may have, are not suitable for its use. 

`cudnn.benchmark = True`, 200 ms inference
<img width="1797" alt="Screen Shot 2020-12-11 at 6 58 20 AM" src="https://user-images.githubusercontent.com/26833433/101918687-7fbddf00-3b7e-11eb-9d99-4158473bdd27.png">

`cudnn.benchmark = False` (default), 12 ms inference
<img width="1797" alt="Screen Shot 2020-12-11 at 6 58 57 AM" src="https://user-images.githubusercontent.com/26833433/101918674-7cc2ee80-3b7e-11eb-8241-479f8a2da6e2.png">


It would be a good idea to raise a bug report on the pytorch repo in any case to let them know of the differing model outputs based on the setting, as this should definitely not be happening.

**Comment by glenn-jocher** [comment]
@wudashuo can you test PR #1672 to verify that this fixes the bug? If it all looks good then I will go ahead and merge it. To clone that PR, you can use this command:

`git clone -b stream https://github.com/ultralytics/yolov5`

**Comment by glenn-jocher** [comment]
@wudashuo PR #1672 is merged, so this fix is in master now. Let me know if this fixes the --save-txt issues you were seeing.

**Comment by wudashuo** [comment]
@glenn-jocher  Sorry for the late reply.
1. Yes I agree with you. It's pytorch's bug, and it's working fine on Linux. Later I will raise a bug report on the pytorch repo.
2. I'm afraid that `--save-txt` bug still exist. I've pulled latest commit. 
First, I test my camera:
`python detect.py --source rtsp://admin:iaac@1234!@192.168.6.42:554 --weights weights\yolov5l.pt --save-txt`
and there's only a empty file named 'admin'
![image](https://user-images.githubusercontent.com/5948604/101971213-ac4f2500-3c6a-11eb-8c60-6f9d596c9fbe.png)
And I also test a public rtsp stream:
`python detect.py --source rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mov --weights weights\yolov5l.pt --save-txt`
This time the empty file's name is `mp4`
![image](https://user-images.githubusercontent.com/5948604/101971266-0e0f8f00-3c6b-11eb-83e7-fc53b5c7108d.png)

**Finally** I noticed that there can't be `: \ / ? * " <> |` in the file name,  and my rtsp link includes `:`.
But these symbols often exist in many rtsp, rtmp or http links, so in these cases we can't open `txt_path` to write results.

**Comment by glenn-jocher** [comment]
@wudashuo ah I see, now I understand. This seems like a Windows path bug, as I tested this in MacOS and it worked fine for the same command (screenshot below). I would have expected pathlib to sort this out, as its supposed to be OS agnostic, but you might need to .replace() those special characters in your filename path for it to work properly on windows. I can't debug this on my end, as I don't have a windows environment available, but why don't you try to debug on your side, and then submit a PR if you find a solution?

EDIT: comparing your results and my results, its clear that the path is becoming interrupted at the ':' character. In linux and macos, the Path(str).name operation successfully creates a filename as shown in my screenshot below. In windows it may need some extra help. Solution is probably to create a generic replace_special_char() function in utils.general.py, and then pass the rtsp filenames through it to turn all the special characters into underscores _.

<img width="1238" alt="Screen Shot 2020-12-11 at 8 05 35 PM" src="https://user-images.githubusercontent.com/26833433/101973343-402ddc00-3bec-11eb-88cf-40c900f50075.png">

**Comment by wudashuo** [comment]
Sure, and again thanks for your amazing work for yolov5!

**Comment by glenn-jocher** [comment]
Maybe something like this could be suitable for utils.general.py. These were the main special characters I could find on my Macbook Pro Spanish keyboard. Got the idea from https://stackoverflow.com/questions/8800815/how-to-replace-all-those-special-characters-with-white-spaces-in-python/40218078

```python
import re

def clean_str(s):
    # Cleans a string by replacing special characters with underscore _
    return re.sub(pattern="[|@#!·$%&()=?¿^*;:,¨´><+]", repl="_", string=s)
```

**Comment by glenn-jocher** [comment]
@wudashuo ok I've implemented string cleaning now in PR #1674.
